### PR TITLE
Cl/replace moment with datefns

### PR DESCRIPTION
**Pull request by clekili**
#### Changes proposed in this pull request:

Replaced moment and most DateUtils functions with date-fns on DateInput component. 

#### Reviewers should focus on:

Dynamic require of locales in dateUtils. I'm not certain that this is a good way of doing this.

Not ready to merge yet, just pushing this PR as a preview.

### CODE REVIEW

File: packages/datetime/package.json [hunk 1]
```diff
@@ -33,7 +33,8 @@
     "classnames": "^2.2",
     "moment": "^2.14.1",
```
Comment (giladgray) [inline packages/datetime/package.json]: can we remove this hopefully-now-obsolete dependency?

File: packages/datetime/src/common/dateUtils.ts [hunk 2]
```diff
@@ -309,3 +310,20 @@ export function stringToMoment(dateString: string, format: DateFormat, locale: s
         return date === undefined ? moment.invalid() : moment(date);
     }
 }
+
+
+export function getLocale(localeString: String): any {
+    if(!localeString) {
+        return undefined;
+    }
+    return require(`date-fns/locale/${localeString}`);
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: ⚠️ probably can't ship this dynamic `require` statement... 
thoughts @adidahiya?

File: packages/datetime/src/common/dateUtils.ts [hunk 3]
```diff
@@ -309,3 +310,20 @@ export function stringToMoment(dateString: string, format: DateFormat, locale: s
...
+}
+
+
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: one blank line please (same on line 314)

File: packages/datetime/src/common/dateUtils.ts [hunk 4]
```diff
@@ -309,3 +310,20 @@ export function stringToMoment(dateString: string, format: DateFormat, locale: s
...
+export function dateToString(date: Date, format: DateFormat, locale: string | undefined) {
+    if (typeof format === "string") {
+        return dateFnFormat(date, format, {locale: getLocale(locale)});
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: it seems like if we dropped support for format strings then users could easily plug in any library like `{ dateToString: (date) => dateFnFormat(date, ...) }`.

this approach might solve the dynamic locale requires too?

File: packages/webpack-build-scripts/index.js [hunk 5]
```diff
@@ -14,6 +14,7 @@ module.exports = {
         "@blueprintjs/labs": "var Blueprint.Labs",
         "@blueprintjs/table": "var Blueprint.Table",
         "classnames": "classNames",
+        "date-fns": "date-fns",
```
Comment (giladgray) [inline packages/webpack-build-scripts/index.js]: `date-fns` is almost certainly not the name of the global variable, as hyphens do not a valid JS identifier make. maybe it's `dateFns`? can you please investigate this?

Comment (clekili) [reply packages/webpack-build-scripts/index.js]: ah you're right it's `dateFns` i'll make the change

**Comment by clekili** [comment]
## Datetime library comparison

### Library Size
moment: 943.58 KB
moment-timezone: 188.62 KB
luxon: 217.4 KB
date-fns: 170.93 KB (when all functions imported)
date-fns: 39.26 KB (when only functions used in datetime package imported)

note: date-fns imports need refactoring for smaller input size importing from root package does load all functions even if some are required. Need to import like `import * as parse from 'date-fns/parse'` instead of `import { parse } from 'date-fns'`

[webpack-bundle-size-analyzer](https://www.npmjs.com/package/webpack-bundle-size-analyzer) was used for size analysis.

### Performance
[https://jsperf.com/momentjs-vs-luxon-vs-date-fns](https://jsperf.com/momentjs-vs-luxon-vs-date-fns)
screen shot 2017-12-22 at 4 06 15 pm

Image above represents tests run on Chrome, performances of libraries may vary in between browsers. The functions used for testing are some of the functions we utilize at @blueprintjs/datetime frequently.

**Comment by clekili** [comment]
It seems like date-fns v2.0.0 is not too far away https://github.com/date-fns/date-fns/milestone/3

@giladgray Similar to what you were mentioning, we may even adopt the pluggable locale model from date-fns v2.0.0 and maybe drop our own `IDateFormatter` ? Here's how it's going to be in date-fns v2.0.0 https://date-fns.org/v2.0.0-alpha.7/docs/Locale

File: packages/datetime/src/dateInput.tsx [hunk 6]
```diff
@@ -320,43 +311,34 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
...
 
     private handleInputFocus = (e: React.FocusEvent<HTMLInputElement>) => {
-        let valueString: string;
-        if (isMomentNull(this.state.value)) {
-            valueString = "";
-        } else {
-            valueString = momentToString(this.state.value, this.props.format, this.props.locale);
-        }
-
+        const valueString = this.getDateString(this.state.value);
```
Comment (cmslewis) [inline packages/datetime/src/dateInput.tsx]: The logic here isn't quite the same as `getDateString`, so we can't replace it wholesale. e.g. we don't want to show out-of-range errors or invalid-date error messages here; those should only appear on blur. Instead, we want to show the original, invalid date in the field on focus.

Are there tests checking this behavior?

File: packages/datetime/src/dateInput.tsx [hunk 7]
```diff
@@ -387,28 +369,28 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
 
     private handleInputBlur = (e: React.FocusEvent<HTMLInputElement>) => {
         const { valueString } = this.state;
-        const value = this.createMoment(valueString);
+        const value = stringToDate(valueString);
         if (
             valueString.length > 0 &&
             valueString !== this.getDateString(this.state.value) &&
-            (!value.isValid() || !this.isMomentInRange(value))
+            (!isValid(value) || !this.isDateInRange(value))
```
Comment (cmslewis) [inline packages/datetime/src/dateInput.tsx]: Can we use `isDateValidAndInRange` here?

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 8]
```diff
@@ -221,4 +221,83 @@ describe("dateUtils", () => {
...
+        it("handles improperly formatted  'MM/D/YYYY'", () => {
+            const date = DateUtils.stringToDate("12/5/2015");
+            expect(toHyphenatedDateString(date)).to.equal("2015-12-05");
+        });
+
+        it("handles improperly formatted  'M/D/YYYY'", () => {
+            const date = DateUtils.stringToDate("2/5/2015");
+            expect(toHyphenatedDateString(date)).to.equal("2015-02-05");
+        });
+    });
```
Comment (cmslewis) [inline packages/datetime/test/common/dateUtilsTests.tsx]: Also handles dates with no hyphens. Would prefer a check for all those weird cases so we know it's intentional.

File: packages/datetime/src/dateRangeInput.tsx [hunk 9]
```diff
@@ -847,19 +835,19 @@ export class DateRangeInput extends AbstractPureComponent<IDateRangeInputProps,
         return boundary === DateRangeBoundary.START ? this.refHandlers.startInputRef : this.refHandlers.endInputRef;
     };
 
-    private getFormattedDateString = (momentDate: moment.Moment, formatOverride?: DateFormat) => {
-        if (isMomentNull(momentDate)) {
+    private getFormattedDateString = (date: Date, formatOverride?: DateFormat) => {
+        if (!date) {
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Use explicit boolean expression

File: packages/datetime/src/dateRangeInput.tsx [hunk 10]
```diff
@@ -957,16 +942,16 @@ export class DateRangeInput extends AbstractPureComponent<IDateRangeInputProps,
...
         const { isInputFocused, hoverString, inputString, selectedValue } = values;
 
-        const boundaryValue = isInputFocused ? this.dateStringToMoment(inputString) : selectedValue;
+        const boundaryValue = isInputFocused ? stringToDate(inputString) : selectedValue;
 
         if (hoverString != null) {
             // don't show an error state while we're hovering over a valid date.
             return false;
-        } else if (isMomentNull(boundaryValue)) {
+        } else if (boundaryValue || boundaryValue == null) {
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Explicit boolean expression only. Also, what's the intended difference in the two checks here?

File: packages/datetime/src/dateRangeInput.tsx [hunk 11]
```diff
@@ -975,16 +960,16 @@ export class DateRangeInput extends AbstractPureComponent<IDateRangeInputProps,
         }
     };
 
-    private isMomentValidAndInRange = (momentDate: moment.Moment) => {
-        return isMomentValidAndInRange(momentDate, this.props.minDate, this.props.maxDate);
-    };
+    private isDateValidAndInRange(value: Date) {
+        return value && isValid(value) && this.isDateInRange(value);
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Explicit boolean expression: `value != null`

File: packages/datetime/src/common/dateUtils.ts [hunk 12]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
-
-export function stringToMoment(dateString: string, format: DateFormat, locale: string | undefined) {
-    if (typeof format === "string") {
-        return moment(dateString, format, locale);
-    } else {
-        const date = format.stringToDate(dateString);
-        return date === undefined ? moment.invalid() : moment(date);
+export function stringToDate(date: string | number | Date) {
+    if (typeof date === "string") {
+        const parseToken = /^(\d{4}|\d{2})-?(\d{1,2})-?(\d{1,2})$/;
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: We can drop the support for optional hyphens if preferred? or we can support 2 digit months and days only on the formats without hyphens. i.e. YYYYMMDD and YYMMDD only. Thoughts?

File: packages/datetime/src/dateInput.tsx (hunk 6 above)
Comment (clekili) [reply packages/datetime/src/dateInput.tsx]: Changed this to use `dateToString` value which shows the original, out of bounds date on focus. But this doesn't handle the invalid dates. Invalid string inputs are attempted to be converted to a date object, which is saved in props. When a string is an invalid date string the prop saved is `Invalid Date` and the string is lost. We can try and change this but it could require us saving the input as a string as well.

Comment (clekili) [reply packages/datetime/src/dateInput.tsx]: Changed this to use `dateToString` value which shows the original, out of bounds date on focus. But this doesn't handle the invalid dates. Invalid string inputs are attempted to be converted to a date object, which is saved in props. When a string is an invalid date string the prop saved is `Invalid Date` and the string is lost. We can try and change this but it could require us saving the input as a string as well.

File: packages/datetime/README.md [hunk 13]
```diff
@@ -8,7 +8,7 @@ This package contains a collection of React components for working with dates an
 
 Two third-party libraries are used in this package to ease the burden of time management:
 
-- [Moment.js](http://momentjs.com/) is the de facto library for manipulating dates in JavaScript.
+- [date-fns](https://date-fns.org) is provides a toolset for manipulating dates in JavaScript.
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/README.md]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 14]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
-}
-
-export function getDatePreviousMonth(date: Date): Date {
-    if (date.getMonth() === Months.JANUARY) {
-        return new Date(date.getFullYear() - 1, Months.DECEMBER);
+    } else if (LOCALE.hasOwnProperty(localeString)) {
+        return LOCALE[localeString];
     } else {
-        return new Date(date.getFullYear(), date.getMonth() - 1);
+        return LOCALE.en;
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 15]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
     }
 }
 
-export function getDateNextMonth(date: Date): Date {
-    if (date.getMonth() === Months.DECEMBER) {
-        return new Date(date.getFullYear() + 1, Months.JANUARY);
+export function dateToString(date: Date, dateFormat: DateFormat, locale = "en") {
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 16]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
+    if (!date) {
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 17]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
+        return format(date, dateFormat, { locale: getLocale(locale) });
     } else {
-        return new Date(date.getFullYear(), date.getMonth() + 1);
+        return dateFormat.dateToString(date);
     }
 }
 
 /**
- * Returns a date string in the provided format localized to the provided locale.
+ * wrapper around date-fn's parse function
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 18]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
-    }
-}
-
-export function stringToMoment(dateString: string, format: DateFormat, locale: string | undefined) {
-    if (typeof format === "string") {
-        return moment(dateString, format, locale);
-    } else {
-        const date = format.stringToDate(dateString);
-        return date === undefined ? moment.invalid() : moment(date);
+export function stringToDate(date: string | number | Date) {
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 19]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
+    if (typeof date === "string") {
+        const parseToken = /^(\d{4}|\d{2})-?(\d{1,2})-?(\d{1,2})$/;
+        const token = parseToken.exec(date.toString());
+        if (token) {
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 20]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
+            const year = token[1];
+            const month = ("0" + token[2]).slice(-2);
+            const day = ("0" + token[3]).slice(-2);
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/dateUtils.ts [hunk 21]
```diff
@@ -189,123 +176,43 @@ export function getDateOnlyWithTime(date: Date): Date {
...
+
+            return parse(`${year}-${month}-${day}`);
+        }
+        return parse(date);
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: Done

File: packages/datetime/src/common/monthAndYear.ts [hunk 22]
```diff
@@ -38,12 +41,12 @@ export class MonthAndYear {
     }
 
     public getPreviousMonth(): MonthAndYear {
-        const previousMonthDate = getDatePreviousMonth(this.date);
+        const previousMonthDate = subMonths(this.date, 1);
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/common/monthAndYear.ts]: Aliased it to `subtractMonths` on import

File: packages/datetime/src/dateInput.tsx [hunk 23]
```diff
@@ -178,21 +175,21 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
     public constructor(props?: IDateInputProps, context?: any) {
         super(props, context);
 
-        const defaultValue = this.props.defaultValue ? fromDateToMoment(this.props.defaultValue) : moment(null);
+        const defaultValue = this.props.defaultValue ? this.props.defaultValue : null;
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/dateInput.tsx]: Done

File: packages/datetime/src/dateInput.tsx [hunk 24]
```diff
@@ -266,31 +263,26 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
...
 
-    private createMoment(valueString: string) {
-        // Locale here used for parsing, does not set the locale on the moment itself
-        return stringToMoment(valueString, this.props.format, this.props.locale);
-    }
-
-    private getDateString = (value: moment.Moment) => {
-        if (isMomentNull(value)) {
+    private getDateString(value: Date) {
+        if (value === null) {
```
[the remark this answers is not in the document]
Comment (clekili) [reply packages/datetime/src/dateInput.tsx]: Done

File: packages/datetime/src/dateInput.tsx (hunk 7 above)
Comment (clekili) [reply packages/datetime/src/dateInput.tsx]: Done

File: packages/datetime/src/dateRangeInput.tsx (hunk 9 above)
Comment (clekili) [reply packages/datetime/src/dateRangeInput.tsx]: Done

File: packages/datetime/src/dateRangeInput.tsx (hunk 10 above)
Comment (clekili) [reply packages/datetime/src/dateRangeInput.tsx]: Changed to `boundaryValue == null`

File: packages/datetime/src/dateRangeInput.tsx (hunk 11 above)
Comment (clekili) [reply packages/datetime/src/dateRangeInput.tsx]: Done

**Comment by adidahiya** [comment]
Please merge develop; your CI build configuration is outdated.

File: packages/datetime/src/common/dateUtils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (cmslewis) [reply packages/datetime/src/common/dateUtils.ts]: @giladgray thoughts?

**Comment by clekili** [comment]
@cmslewis changed the date parsing to only accept following formats:
`YYYY-MM-DD`, `YYYY-MM-D`, `YYYY-M-DD`, `YY-MM-DD`, `YY-M-D`, `YYYYMMDD`, `YYMMDD`
Going to add tests around this.

Another thing to do is to add tests for on focus of `DateInput` to check it indeed renders the actual out of range date instead of the string 'Out of Range'. I don't yet have a good solution to handle the invalid dates same way though, let me know if you have any ideas about this. The solution I'm thinking of involves changing the way we store the inputs which I'm not sure if it's worth that.

File: packages/datetime/src/common/dateUtils.ts [hunk 25]
```diff
@@ -4,27 +4,26 @@
...
 
-import * as moment from "moment";
+import * as format from "date-fns/format";
+import * as isSameDay from "date-fns/is_same_day";
+import * as parse from "date-fns/parse";
+
+// Import a subset of locales to reduce the bundle size of consuming apps
+// Following locales have been selected according to their popularity
+const LOCALE: { [index: string]: any } = {
+    de: require("date-fns/locale/de"),
```
Comment (adidahiya) [inline packages/datetime/src/common/dateUtils.ts]: these are static requires. currently, the locales get imported as soon as `@blueprintjs/datetime` is imported, which isn't what we want. users should be able to supply their own list of locales from `date-fns` if we end up going this route.

Comment (clekili) [reply packages/datetime/src/common/dateUtils.ts]: These static requires are intended to be a workaround until `date-fns` releases a `v2.0` in which it's going to be possible to support user specified locales.

There also are plans for `date-fns` to support timezones:
 - https://github.com/date-fns/date-fns/issues/180
 - https://github.com/date-fns/date-fns/issues/376

I think date-fns have good potential, but it needs to improve and is under active development.

Another lib that might be worth exploring : https://github.com/smallwins/spacetime

**Comment by adidahiya** [comment]
@clekili I agree that date-fns has potential, but we need to ship 2.0 very soon and can't wait on date-fns to do all those things. We might need to stick with moment for now.

Why spacetime over luxon? (btw I don't think the jsperf benchmarks are very relevant since our datetime component interactions are bounded by human input)

**Comment by clekili** [comment]
@adidahiya I haven't been able to do an in-depth comparison analysis between spacetime and luxon, it's just a lib I've heard of that might be worth a look into. 

Do we have a timeline for shipping 2.0? I do have some impl work to attend to but I also want to get this done.

**Comment by adidahiya** [comment]
timeline: this week. hoping to cut an RC with all the expected breaking changes ASAP
